### ISSUE TITLE
Data differs when Bytedelta used with zstd with chunk size % type size != 0

### ISSUE THREAD

**Issue by martaiborra** [Blosc/c-blosc2#513]
**Describe the bug**
When compressing an arange chunk of size 50 * 4,  type size 7 (real type size 4), codec ZSTD and filters shuffle + byte delta the lasts elements of the chunk are not retrieved correctly.

**To Reproduce**
The following code reproduces the problem. It creates a sframe with one chunk. 

```
/*
  Copyright (c) 2021  The Blosc Development Team <blosc@blosc.org>
  https://blosc.org
  License: BSD 3-Clause (see LICENSE.txt)

 Example program demonstrating use of the Blosc filter from C code.

 To compile this program:

 $ gcc sframe_simple.c -o sframe_simple -lblosc2

 To run:

 $ ./sframe_simple
*/

#include <stdio.h>
#include <blosc2.h>
#include "blosc2/filters-registry.h"

#define KB  1024.
#define MB  (1024*KB)
#define GB  (1024*MB)

#define CHUNKSIZE (50)
#define NCHUNKS 1
#define NTHREADS 2


int main(void) {
  blosc2_init();
  static int32_t data[CHUNKSIZE];
  static int32_t data_dest[CHUNKSIZE];
  int32_t isize = CHUNKSIZE * sizeof(int32_t);
  int dsize;
  int64_t nbytes, cbytes;
  blosc2_cparams cparams = BLOSC2_CPARAMS_DEFAULTS;
  cparams.compcode = BLOSC_ZSTD;
  cparams.filters[BLOSC2_MAX_FILTERS - 2] = BLOSC_SHUFFLE;

  cparams.filters[BLOSC2_MAX_FILTERS - 1] = BLOSC_FILTER_BYTEDELTA;
  cparams.filters_meta[BLOSC2_MAX_FILTERS - 1] = 7;

  blosc2_dparams dparams = BLOSC2_DPARAMS_DEFAULTS;
  blosc2_schunk* schunk;
  int i, nchunk;
  blosc_timestamp_t last, current;
  double ttotal;

  printf("Blosc version info: %s (%s)\n", blosc2_get_version_string(), BLOSC2_VERSION_DATE);
  /* Create a super-chunk container */
  cparams.typesize = 7;
  cparams.nthreads = NTHREADS;
  dparams.nthreads = NTHREADS;
  blosc2_storage storage = {false, "dir1.b2frame", .cparams=&cparams, .dparams=&dparams};
  /* Remove directory */
  blosc2_remove_dir(storage.urlpath);

  schunk = blosc2_schunk_new(&storage);

  blosc_set_timestamp(&last);
  for (nchunk = 0; nchunk < NCHUNKS; nchunk++) {
    for (i = 0; i < CHUNKSIZE; i++) {
      data[i] = i;
    }
    int64_t nchunks = blosc2_schunk_append_buffer(schunk, data, isize);
    if (nchunks != nchunk + 1) {
      printf("Unexpected nchunks!");
      return -1;
    }
  }
  /* Gather some info */
  nbytes = schunk->nbytes;
  cbytes = schunk->cbytes;
  blosc_set_timestamp(&current);
  ttotal = blosc_elapsed_secs(last, current);
  printf("Compression ratio: %.1f MB -> %.1f MB (%.1fx)\n",
         (double)nbytes / MB, (double)cbytes / MB, (1. * (double)nbytes) / (double)cbytes);
  printf("Compression time: %.3g s, %.1f MB/s\n",
         ttotal, (double)nbytes / (ttotal * MB));

  /* Retrieve and decompress the chunks (0-based count) */
  blosc_set_timestamp(&last);
  for (nchunk = NCHUNKS-1; nchunk >= 0; nchunk--) {
    dsize = blosc2_schunk_decompress_chunk(schunk, nchunk, data_dest, isize);
    if (dsize < 0) {
      printf("Decompression error.  Error code: %d\n", dsize);
      return dsize;
    }
  }
  blosc_set_timestamp(&current);
  ttotal = blosc_elapsed_secs(last, current);
  printf("Decompression time: %.3g s, %.1f MB/s\n",
         ttotal, (double)nbytes / (ttotal * MB));

  blosc2_schunk_decompress_chunk(schunk, 0, data_dest, isize);
  for (i = 0; i < CHUNKSIZE; i++) {
    if (data_dest[i] != i) {
      printf("Decompressed data differs from original %d, %d!\n",
             i, data_dest[i]);
      return -1;
    }
  }

  printf("Successful roundtrip data <-> schunk !\n");

  /* Remove directory */
  blosc2_remove_dir(storage.urlpath);
  /* Free resources */
  /* Destroy the super-chunk */
  blosc2_schunk_free(schunk);
  blosc2_destroy();
  return 0;
}

```

**Expected behavior**
When checking the retrieved data it should end with a `"Successful roundtrip data <-> schunk !"` message. Instead:

```
Blosc version info: 2.9.3.dev ($Date:: 2023-05-18 #$)
Compression ratio: 0.0 MB -> 0.0 MB (0.9x)
Compression time: 0.00176 s, 0.1 MB/s
Decompression time: 0.000379 s, 0.5 MB/s
Decompressed data differs from original 49, 0!
```

**Additional context**
This works fine if the codec used is BLOSCLZ..

**Comment by froody** [comment]
I'm hitting this too!

**Comment by froody** [comment]
Actually I think my issue is separate to yours. You're using a `typesize` of 7 with a chunksize of 50 bytes, I'm new to blosc but that seems particularly error prone. If you add logging to `bytedelta_backward` you'll see it being invoked with `length = 4, meta=7`, which results in a stream_len of 0. I'll create a separate issue for my problem.

**Comment by FrancescAlted** [comment]
Fixed in #532 .  Thanks @froody !
